pkg/agent/controller/l7visibility/l7visibility_controller_test.go

```diff
@@ -0,0 +1,546 @@
+	interfaces := []*interfacestore.InterfaceConfig{
+		podInterface1,
+		podInterface2,
+	}
+	testcases := []struct {
+		name                      string
+		addedPod                  *v1.Pod
+		expectedPodToDirectionMap map[string]v1alpha2.Direction
+		expectedCalls             func(mockOFClient *openflowtest.MockClient)
+		expectedError             string
+	}{
+		{
+			name:     "Add pod with correct annotations",
+			addedPod: pod1,
+			expectedPodToDirectionMap: map[string]v1alpha2.Direction{
+				pod1NN: v1alpha2.DirectionIngress,
+			},
+			expectedCalls: func(mockOFClient *openflowtest.MockClient) {
+				mockOFClient.EXPECT().InstallTrafficControlMarkFlows(fmt.Sprintf("tcl7:%s", pod1NN), []uint32{uint32(podInterface1.OFPort)}, targetPort, directionIngress, actionMirror, types.TrafficControlFlowPriorityLow)
+			},
+		}, {
+			name:                      "Add pod with incorrect annotations",
+			addedPod:                  pod2,
+			expectedPodToDirectionMap: map[string]v1alpha2.Direction{},
+			expectedCalls:             func(mockOFClient *openflowtest.MockClient) {},
+			expectedError:             fmt.Sprintf("annotation key %s can only have values (Ingress/Egress/Both)", types.L7VisibilityAnnotationKey),
+		},
+	}
+	for _, tt := range testcases {
+		t.Run(tt.name, func(t *testing.T) {
+			c := newFakeControllerAndWatcher(t, []runtime.Object{tt.addedPod, testNS1}, interfaces)
+			stopCh := make(chan struct{})
+			defer close(stopCh)
+
+			c.startInformers(stopCh)
+			assert.Eventuallyf(t, func() bool {
+				ns, _ := c.namespaceLister.List(labels.Everything())
+				return len(c.localPodInformer.GetIndexer().List()) == 1 && len(ns) == 1
+			}, 1*time.Second, 10*time.Millisecond, "Pod should be added to Informers")
+			waitEvents(t, 1, c)
+			item, _ := c.queue.Get()
+			tt.expectedCalls(c.mockOFClient)
+			err := c.syncPod(item.(string))
+			if tt.expectedError != "" {
+				assert.ErrorContains(t, err, tt.expectedError)
+			} else {
+				assert.Equal(t, tt.expectedPodToDirectionMap, c.podToDirectionMap)
+			}
+			c.queue.Done(item)
+		})
+	}
+}
+
+func TestPodUpdate(t *testing.T) {
+	var targetPort uint32
+	testNS1 := newNamespaceObject("test-ns1", annotationsEmpty)
+	pod1 := newPodObject("test-pod1", "test-ns1", annotationsDifferent)
+	pod2 := newPodObject("test-pod2", "test-ns1", annotationsIncorrect)
+	pod3 := newPodObject("test-pod3", "test-ns1", annotationsEmpty)
+	pod1NN := k8s.NamespacedName(pod1.Namespace, pod1.Name)
+	pod2NN := k8s.NamespacedName(pod2.Namespace, pod2.Name)
+	pod3NN := k8s.NamespacedName(pod3.Namespace, pod3.Name)
+	podInterface1 := newPodInterface(pod1.Name, pod1.Namespace, int32(pod1OFPort))
+	podInterface2 := newPodInterface(pod2.Name, pod2.Namespace, int32(pod2OFPort))
+	podInterface3 := newPodInterface(pod3.Name, pod3.Namespace, int32(pod3OFPort))
+	interfaces := []*interfacestore.InterfaceConfig{
+		podInterface1,
+		podInterface2,
+		podInterface3,
+	}
+	testcases := []struct {
```

Comment:
I notice we don't have a test for a Pod update with a direction change (e.g., ingress -> egress, or ingress -> both, or both -> egress)

Reply:
I don't think we require all scenarios , we can have either of the above and that should cover all test cases